src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItemSource.cs

```diff
@@ -0,0 +1,281 @@
+        /// events and expect that.
+        /// </summary>
+        private readonly BulkObservableCollectionWithInit<BaseItem> _items = new();
+
+        /// <summary>
+        /// Gate to guard mutation of <see cref="_cancellationTokenSource"/> and <see cref="_resettableDelay"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        private CancellationTokenSource? _cancellationTokenSource;
+        private ResettableDelay? _resettableDelay;
+
+        public SourceGeneratedFileItemSource(SourceGeneratorItem parentGeneratorItem, Workspace workspace, IAsynchronousOperationListener asyncListener, IThreadingContext threadingContext)
+        {
+            _parentGeneratorItem = parentGeneratorItem;
+            _workspace = workspace;
+            _asyncListener = asyncListener;
+            _threadingContext = threadingContext;
+        }
+
+        public object SourceItem => _parentGeneratorItem;
+
+        public bool HasItems
+        {
+            get
+            {
+                // Since we are expensive to compute, always say we have items.
+                return true;
+            }
+        }
+
+        public IEnumerable Items => _items;
+
+        private async Task UpdateSourceGeneratedFileItemsAsync(Solution solution, CancellationToken cancellationToken)
+        {
+            var project = solution.GetProject(_parentGeneratorItem.ProjectId);
+
+            if (project == null)
+            {
+                return;
+            }
+
+            var sourceGeneratedDocuments = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
+            var sourceGeneratedDocumentsForGeneratorById = sourceGeneratedDocuments.Where(d => d.SourceGenerator == _parentGeneratorItem.Generator).ToDictionary(d => d.Id);
+
+            // We must update the list on the UI thread, since the WPF elements bound to our list expect that
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            try
+            {
+                // We're going to incrementally update our items, ensuring we keep the object identity for things we didn't touch.
+                _items.BeginBulkOperation();
+
+                // Do we already have a "no files" placeholder item?
+                if (_items.Count == 1 && _items[0] is NoSourceGeneratedFilesPlaceholderItem)
+                {
+                    // We do -- if we have no items, we're done, since the placeholder is all that needs to be there;
+                    // otherwise remove it since we have real files now
+                    if (sourceGeneratedDocumentsForGeneratorById.Count == 0)
+                    {
+                        return;
+                    }
+                    else
+                    {
+                        _items.RemoveAt(0);
+                    }
+                }
+
+                for (int i = 0; i < _items.Count; i++)
```

Comment:
My understanding is the solution explorer is also doing similar comparisons of identity to ensure that nodes maintain "lifetime". Consider the case where you have a node selected (because you're about to hit enter to navigate to it). In that case, if we were to refresh right before that, we don't want to lose the selection or have it select something else.

I probably should give a comment that's why we're doing that.

Reply:
The pithy, one-line comment that was indecipherable has been expanded. :smile: